diff-tests/core/model/src/model.ts

```diff
@@ -0,0 +1,671 @@
+    this.endBlockMaturation();
+  };
+
+  delegate = (val: Validator, amt: number) => {
+    this.delegatorTokens -= amt;
+    this.tokens[val] += amt;
+    this.delegation[val] += amt;
+  };
+
+  undelegate = (val: Validator, amt: number) => {
+    if (this.delegation[val] < amt) {
+      this.m.events.push(Event.INSUFFICIENT_SHARES);
+      return;
+    }
+    this.tokens[val] -= amt;
+    this.delegation[val] -= amt;
+    const und: Undelegation = {
+      val: val,
+      creationHeight: this.m.h[P],
+      completionTime: this.m.t[P] + UNBONDING_SECONDS_P,
+      balance: amt,
+      initialBalance: amt,
+      onHold: true,
+      opID: this.opID,
+      willBeProcessedByStakingModule: true,
+    };
+    this.undelegationQ.push(und);
+    this.m.ccvP.afterUnbondingInitiated(this.opID);
+    this.opID += 1;
+  };
+
+  slash = (val: Validator, infractionHeight: number) => {
+    const valid = (e: Undelegation): boolean =>
+      e.val === val &&
+      infractionHeight <= e.creationHeight &&
+      (this.m.t[P] < e.completionTime || e.onHold);
+    const ubds: Undelegation[] = this.undelegationQ.filter(valid);
+    if (infractionHeight < this.m.h[P]) {
+      ubds.forEach(() => {
+        this.m.events.push(Event.SLASH_UNDEL);
+      });
+    }
+  };
+
+  jailUntil = (val: Validator, timestamp: number) => {
+    this.jailed[val] = timestamp;
+    this.m.events.push(Event.JAIL);
+  };
+
+  unbondingCanComplete = (opID: number) => {
+    {
+      const e = _.find(this.validatorQ, (e) => e.opID === opID);
+      if (e) {
+        e.onHold = false;
+        this.m.events.push(Event.SET_UNVAL_HOLD_FALSE);
+        return;
+      }
+    }
+    const e = _.find(this.undelegationQ, (e) => e.opID === opID);
+    if (e) {
+      if (e.completionTime <= this.m.t[P]) {
+        this.delegatorTokens += e.balance;
+        this.undelegationQ = this.undelegationQ.filter((x) => x !== e);
+        this.m.events.push(Event.COMPLETE_UNDEL_NOW);
+      } else {
+        e.onHold = false;
+        this.m.events.push(Event.SET_UNDEL_HOLD_FALSE);
+      }
+    }
+  };
+
+  valUpdates = () => {
+    return _.clone(this.changes);
+  };
+}
+
+class CCVProvider {
+  m;
+  // height of onChanOpenConfirm event
+  initialHeight: number;
+  // next id to use
+  vscID: number;
+  // map ids to height of sending
+  // used to calculate infraction height from consumer initiated slashing
+  vscIDtoH: Record<number, number>;
+  // map ids to unbonding operation ids
+  // used to mature unbonding operations when receiving maturity packets
+  vscIDtoOpIDs: Map<number, number[]>;
+  // validators who have been slashed since last VSC
+  downtimeSlashAcks: Validator[];
+  // is a validator tombstoned?
+  tombstoned: boolean[];
+  // unbonding operations to be completed in EndBlock
+  matureUnbondingOps: number[];
+
+  constructor(model: Model, { ccvP }: ModelInitState) {
+    this.m = model;
+    Object.assign(this, ccvP);
+  }
+
+  endBlock = () => {
```

Comment:
Split in two, similarly to the spec, i.e., `EndBlockCIS` and `EndBlockVSU`. Furthermore, I think it could be a nice idea to use the tags from the spec, e.g., `[CCV-PCF-EBLOCK-CIS.1]`.